File: e2e/cypress-javascript/interactions/steps/TaskTableStep.js
const { step } = require('@/decorators/step-decorator');
const { MainPage } = require('@/interactions/pages/MainPage');
const { CreateTaskForm } = require('@/interactions/pages/CreateTaskForm');
const { CreateTaskStep } = require('@/interactions/steps/CreateTaskStep');
const { EditTaskStep } = require('@/interactions/steps/EditTaskStep');
const { TaskInfoModal } = require('@/interactions/pages/TaskInfoModal');
const { EditTaskForm } = require('@/interactions/pages/EditTaskForm');

class TaskTableStep {
  constructor() {
    this.mainPage = new MainPage();
    this.createTaskForm = new CreateTaskForm();
    this.taskInfoModal = new TaskInfoModal();
    this.editTaskForm = new EditTaskForm();
    this.createTask = new CreateTaskStep();
    this.editTask = new EditTaskStep();

    this.openCreateTaskForm = step('Open create task form', this._openCreateTaskForm.bind(this));
    this.openTaskInfoForm = step("Open task info for task '{title}'", this._openTaskInfoForm.bind(this));
    this.openEditTaskForm = step("Open edit form for task '{title}'", this._openEditTaskForm.bind(this));
    this.deleteTask = step("Delete task '{title}'", this._deleteTask.bind(this));
  }

  _openCreateTaskForm() {
    this.mainPage.createTaskButton.click();
    this.createTaskForm.submitButton.should('be.visible');
    return this.createTask;
  }

  _openTaskInfoForm(title) {
    this._resolveTaskId(title).then((id) => {
      this.mainPage.taskInfoButtonById(id).click();
      this.taskInfoModal.title.should('be.visible');
    });
  }

  _openEditTaskForm(title) {
    this._resolveTaskId(title).then((id) => {
      this.mainPage.taskEditButtonById(id).click();
      this.editTaskForm.titleInput.should('have.value', title);
    });
    return this.editTask;
  }

  _deleteTask(title) {
    this._resolveTaskId(title).then((id) => {
      this.mainPage.taskDeleteButtonById(id).click();
      this.mainPage.taskDeleteButtonById(id).should('not.exist');
    });
  }

  _resolveTaskId(title) {
    return this.mainPage.taskTitleByTitle(title)
      .invoke('attr', 'data-testid')
      .then((dataTestId) => dataTestId.replace('task-title-', ''));
  }
}

module.exports = { TaskTableStep };

File: e2e/capybara-ruby/interactions/steps/task_table_step.rb
# frozen_string_literal: true

require_relative '../../decorators/step_decorator'
require_relative '../../support/capybara_helpers'
require_relative '../pages/main_page'
require_relative '../pages/create_task_form'
require_relative '../pages/task_info_modal'
require_relative '../pages/edit_task_form'
require_relative 'create_task_step'
require_relative 'edit_task_step'

class TaskTableStep
  extend StepDecorator::Methods
  include CapybaraHelpers

  attr_reader :create_task, :edit_task

  def initialize
    @main_page = MainPage.new
    @create_task_form = CreateTaskForm.new
    @task_info_modal = TaskInfoModal.new
    @edit_task_form = EditTaskForm.new
    @create_task = CreateTaskStep.new
    @edit_task = EditTaskStep.new
  end

  step 'Open create task form'
  def open_create_task_form
    @main_page.create_task_button.click
    expect(@create_task_form.submit_button).to be_visible
    @create_task
  end

  step "Open task info for task '{title}'"
  def open_task_info_form(title)
    id = resolve_task_id(title)
    @main_page.task_info_button_by_id(id).click
    expect(@task_info_modal.title).to be_visible
  end

  step "Open edit form for task '{title}'"
  def open_edit_task_form(title)
    id = resolve_task_id(title)
    @main_page.task_edit_button_by_id(id).click
    expect(@edit_task_form.title_input.value).to eq(title)
    @edit_task
  end

  step "Delete task '{title}'"
  def delete_task(title)
    id = resolve_task_id(title)
    @main_page.task_delete_button_by_id(id).click
    expect(page).to have_no_css("[data-testid=\"delete-button-#{id}\"]")
  end

  private

  def resolve_task_id(title)
    @main_page.task_title_by_title(title)['data-testid'].delete_prefix('task-title-')
  end
end
